File: frontend/src/types/cluster.ts
/**
 * Challenge cluster types — aligned with backend app/schemas/cluster.py
 */

export type ClusterStatus = "SUGGESTED" | "VALIDATED" | "REJECTED";

export interface ClusterMemberOut {
  challenge_id: number;
  title: string;
  location: string;
  category: string;
  severity: string;
  status: string;
  similarity_score: number | null;
}

export interface ClusterResponse {
  id: number;
  code: string;
  title: string;
  description: string;
  status: ClusterStatus;
  rationale: string;
  created_by: number;
  created_at: string;
  updated_at: string;
  members: ClusterMemberOut[];
}

export interface ClusterCreate {
  title: string;
  description: string;
  rationale: string;
  challenge_ids: number[];
}

export interface SuggestedCluster {
  title: string;
  code: string;
  description: string;
  rationale: string;
  member_count: number;
  locations: string[];
  categories: string[];
  average_similarity: number;
  member_challenge_ids: number[];
  member_titles: string[];
}

export interface ClusterSuggestResponse {
  suggested_clusters: SuggestedCluster[];
  methodology_note: string;
}

File: backend/app/schemas/cluster.py
"""Challenge cluster API schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ClusterStatus = Literal[
    "SUGGESTED",
    "VALIDATED",
    "REJECTED",
]


class ClusterMemberOut(BaseModel):
    """A challenge linked to a cluster, with the evidence captured."""

    challenge_id: int
    title: str
    location: str
    category: str
    severity: str
    status: str
    similarity_score: float | None = None


class ClusterCreate(BaseModel):
    """Create a cluster with explicit member challenge IDs."""

    title: str = Field(min_length=3, max_length=200)
    description: str = Field(min_length=10)
    rationale: str = Field(default="")
    challenge_ids: list[int] = Field(min_length=2)


class ClusterResponse(BaseModel):
    """Cluster as returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    title: str
    description: str
    status: ClusterStatus
    rationale: str
    created_by: int
    created_at: datetime
    updated_at: datetime
    members: list[ClusterMemberOut] = []


class SuggestedCluster(BaseModel):
    """One candidate cluster produced by the suggestion engine."""

    title: str
    code: str
    description: str
    rationale: str
    member_count: int
    locations: list[str]
    categories: list[str]
    average_similarity: float
    member_challenge_ids: list[int]
    member_titles: list[str]


class ClusterSuggestResponse(BaseModel):
    """Candidate clusters for the current set of challenges."""

    suggested_clusters: list[SuggestedCluster]
    methodology_note: str = (
        "Prototype clustering: challenges are grouped by shared category, "
        "nearby location, and text similarity. Suggestions require human "
        "validation before they become systemic challenges."
    )